File: packages/runtime/src/h.mjs
import { withoutNulls } from "./utils/array.mjs"

export const DOM_TYPES={
    TEXT:'text',
    ELEMENT:'element',
    FRAGMENT: 'fragment'
}

export const h = (tag,props={},children=[])=>{
    return {
        tag,
        props,
        children: mapTextNodes(withoutNulls(children)),
        type:DOM_TYPES.ELEMENT
    }
}

const mapTextNodes = (children) =>{
  return children.map((child) => typeof child === 'string' ? hString(child) : child)
}
export const hString = (str) =>{
  return { type:DOM_TYPES.TEXT, value:str }
}

export const hFragment = (vNodes)=>{
  return {
    type: DOM_TYPES.FRAGMENT,
    children: mapTextNodes(withoutNulls(vNodes))
  }
}

File: internals/h.go
package internals

import (
	"go-fe-fwk/types"
	"reflect"
)

func H(tag string, props types.Props, children []any) types.Vdom {
	newChildren := []any{}
	for i := range children {
		if reflect.TypeOf(children[i]) == reflect.TypeOf("") {
			newChildren = append(newChildren, Hstring(children[i].(string)))
		} else {
			newChildren = append(newChildren, children[i])
		}
	}
	return types.Vdom{
		Tag:      tag,
		Props:    props,
		Children: newChildren,
		Type:     types.DOM_TYPES["ELEMENT"],
	}
}

func Hstring(value string) types.StringDom {
	return types.StringDom{
		Type:  types.DOM_TYPES["TEXT"],
		Value: value,
	}
}
